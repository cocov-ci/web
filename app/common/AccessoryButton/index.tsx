import classNames from 'classnames'
import Image from 'next/image'
import Link from 'next/link'
import React, { HTMLAttributeAnchorTarget } from 'react'

import Url from 'types/Url'

import styles from './AccessoryButton.module.scss'

type AccessoryButtonProps = {
  children?: React.ReactNode
  className?: string
  kind: 'squared' | 'round' | 'squared-muted'
  onClick?: (ev: React.MouseEvent) => void
  title?: string
  href?: Url
  disabled?: boolean
  hrefTarget?: HTMLAttributeAnchorTarget
}

const AccessoryButton = ({
  children,
  className,
  kind,
  onClick,
  title,
  href,
  hrefTarget,
  disabled,
}: AccessoryButtonProps) => {
  const baseProps = {
    className: classNames(
      styles.base,
      className,
      ...kind.split('-').map(i => styles[i]),
      {
        [styles.disabled]: disabled,
      },
    ),
    title: title,
  }

  if (href) {
    return (
      <Link href={href} target={hrefTarget}>
        <div {...baseProps}>{children}</div>
      </Link>
    )
  }

  return (
    <button {...baseProps} onClick={ev => onClick && onClick(ev)}>
      {children}
    </button>
  )
}

export const CopyIcon = () => (
  <Image alt="" height={13} src="/icons/copy.svg" width={13} />
)

export const CheckIcon = () => (
  <Image alt="" height={11} src="/icons/check.svg" width={16} />
)

export const HelpIcon = () => (
  <Image alt="" height={10.5} src="/icons/help.svg" width={5.83} />
)

export default AccessoryButton
