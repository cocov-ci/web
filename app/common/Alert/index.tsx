import classNames from 'classnames'
import { LucideIcon } from 'lucide-react'
import React from 'react'

import styles from './Alert.module.scss'

type BoxProps = {
  children?: React.ReactNode
  className?: string
  title?: string
  description?: string
  icon?: LucideIcon
}

const Alert = ({ className, icon, title, description }: BoxProps) => {
  const Icon = icon as LucideIcon

  return (
    <div className={classNames(styles.alert, className)}>
      <div className={styles.box}>
        {icon && <Icon className={styles.icon} size={40} />}
        <div className={styles.content}>
          {title && <h2>{title}</h2>}
          {description && <h3>{description}</h3>}
        </div>
      </div>
    </div>
  )
}

export default Alert
