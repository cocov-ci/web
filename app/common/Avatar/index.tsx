import classNames from 'classnames'
import React from 'react'

import styles from './Avatar.module.scss'

type AvatarProps = {
  className?: string
  avatarURL: string
  size: string
}

const Avatar = ({ avatarURL, className, size }: AvatarProps) => {
  return (
    // eslint-disable-next-line @next/next/no-img-element
    <img
      alt=""
      aria-hidden={true}
      className={classNames(styles.base, className)}
      src={avatarURL}
      style={{
        height: size,
        width: size,
      }}
    />
  )
}

export default Avatar
