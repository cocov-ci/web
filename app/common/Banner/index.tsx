import classNames from 'classnames'
import { LucideIcon, XIcon } from 'lucide-react'
import React from 'react'

import styles from './Banner.module.scss'

export type BannerVariationProps = 'success' | 'warning' | 'info' | 'neutral'

export type BannerProps = {
  className?: string
  variation?: BannerVariationProps
  children: React.ReactNode
  icon?: LucideIcon
  onClose?: () => void
  open?: boolean
}

const Banner = ({
  className,
  variation = 'neutral',
  icon,
  children,
  onClose,
  open,
}: BannerProps) => {
  const IconComponent = icon

  return (
    <div className={styles.wrapper}>
      <div
        className={classNames(styles.faces, {
          [styles.open]: open,
        })}
      >
        <div className={classNames(styles.base, className)}>
          <div className={styles.shadow} />
          <div className={classNames(styles.innerBase, styles[variation])}>
            <div>
              {IconComponent && (
                <div className={styles.icon}>
                  <IconComponent size={20} />
                </div>
              )}
              <div className={styles.content}>{children}</div>
              <div className={styles.close}>
                <button onClick={onClose}>
                  <XIcon size={20} />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  )
}

export default Banner
