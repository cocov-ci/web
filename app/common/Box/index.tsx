import classNames from 'classnames'
import React from 'react'

import styles from './Box.module.scss'

type BoxProps = {
  children?: React.ReactNode
  className?: string
  gutterBottom?: boolean
  backgroundTextSmall?: string
  backgroundTextLarge?: string
}

const Box = ({
  children,
  gutterBottom,
  backgroundTextLarge,
  backgroundTextSmall,
  className,
}: BoxProps) => {
  const backgroundText = backgroundTextLarge || backgroundTextSmall

  return (
    <div
      className={classNames(styles.box, className, {
        [styles.gutterBottom]: gutterBottom,
      })}
    >
      {backgroundText && (
        <span
          className={classNames(styles.backgroundText, {
            [styles.small]: Boolean(backgroundTextSmall),
          })}
        >
          {backgroundText}
        </span>
      )}
      <div className={styles.content}>{children}</div>
    </div>
  )
}

export default Box
