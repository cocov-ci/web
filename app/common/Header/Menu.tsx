'use client'

import Link from 'next/link'
import { useEffect, useState } from 'react'

import useAuth from 'hooks/useAuth'
import { UserProps } from 'types/User'

import styles from './Header.module.scss'

const Menu = () => {
  const [loggedUser, setLoggedUser] = useState<UserProps>()
  const { logout, user } = useAuth()

  useEffect(() => {
    setLoggedUser(user)
  }, [user])

  if (!loggedUser) return null

  return (
    <nav>
      <ul className={styles.menu}>
        <li>
          <Link href="/">Repositories</Link>
        </li>
        {loggedUser?.isAdmin && (
          <li>
            <Link href="#">Adminland</Link>
          </li>
        )}
        <li>
          <p onClick={() => logout()}>Sign Out</p>
        </li>
      </ul>
    </nav>
  )
}

export default Menu
