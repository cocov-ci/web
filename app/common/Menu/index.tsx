'use client'

import classNames from 'classnames'
import { LucideIcon } from 'lucide-react'
import Link from 'next/link'
import React, { forwardRef } from 'react'

import { satoshi } from 'utils/fonts'

import styles from './Menu.module.scss'

interface MenuItemProps {
  label: string
  icon?: LucideIcon
  danger?: boolean
  onClick?: () => void
  href?: string
}

export const MenuItem = ({
  label,
  icon,
  danger,
  onClick,
  href,
}: MenuItemProps) => {
  const IconComponent = icon

  const contents = (
    <div
      className={classNames(styles.item, {
        [styles.danger]: danger,
      })}
    >
      <span className={classNames(styles.label, satoshi.className)}>
        {label}
      </span>
      {IconComponent && <IconComponent size={16} />}
    </div>
  )

  if (onClick) {
    return (
      <a href="javascript:void(0)" onClick={onClick}>
        {contents}
      </a>
    )
  } else if (href) {
    return <Link href={href}>{contents}</Link>
  } else {
    return contents
  }
}

interface MenuProps {
  children?: React.ReactNode
  visible?: boolean
  className?: string
}

export const MenuContainer = forwardRef<HTMLDivElement, MenuProps>(
  ({ children, visible, className }: MenuProps, ref) => (
    <div
      className={classNames(styles.base, className, {
        [styles.visible]: visible,
      })}
      ref={ref}
    >
      <div className={styles.background} />
      <div className={styles.content}>{children}</div>
    </div>
  ),
)
