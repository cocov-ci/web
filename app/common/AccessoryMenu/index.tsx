import classNames from 'classnames'
import { LucideIcon, MoreVertical } from 'lucide-react'
import React, { useRef, useState } from 'react'

import useOnClickOutside from 'hooks/useOnClickOutside'

import AccessoryButton from '../AccessoryButton'
import { MenuContainer } from '../Menu'

import styles from './AccessoryMenu.module.scss'

interface AccessoryMenuProps {
  className?: string
  icon?: LucideIcon
  children: React.ReactNode
  open?: boolean
  muted?: boolean
}

const AccessoryMenu = ({
  className,
  icon,
  open,
  children,
  muted,
}: AccessoryMenuProps) => {
  const Icon = icon || MoreVertical
  const menuContainerRef = useRef<HTMLDivElement>(null)
  const [visible, setVisible] = useState(open)
  useOnClickOutside(menuContainerRef, () => setVisible(false))

  return (
    <div className={classNames(styles.base, className)}>
      <AccessoryButton
        className={styles.button}
        kind={muted ? 'squared-muted' : 'squared'}
        onClick={() => setVisible(!visible)}
      >
        <Icon size={20} />
      </AccessoryButton>
      <MenuContainer
        className={styles.container}
        ref={menuContainerRef}
        visible={visible}
      >
        {children}
      </MenuContainer>
    </div>
  )
}

export default AccessoryMenu
