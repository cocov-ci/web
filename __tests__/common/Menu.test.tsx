import { fireEvent, render, screen } from '@testing-library/react'
import {
  LucideLogOut,
  LucideRefreshCw,
  LucideTrash2,
  LucideUserX,
} from 'lucide-react'

import { MenuContainer, MenuItem } from 'app/common/Menu'

import '@testing-library/jest-dom'

describe('common/Menu', () => {
  it('renders Menu', () => {
    const { container } = render(
      <MenuContainer>
        <MenuItem href="." icon={LucideUserX} label="Revoke Admin" />
        <MenuItem
          icon={LucideRefreshCw}
          label="Sync Permissions"
          onClick={() => jest.fn()}
        />
        <MenuItem icon={LucideLogOut} label="Force Logout" />
        <MenuItem danger icon={LucideTrash2} label="Delete..." />
      </MenuContainer>,
    )

    expect(container.querySelectorAll('.content a').length).toEqual(2)
    expect(container.querySelectorAll('.content .item').length).toEqual(4)
  })

  it('renders onClick function', () => {
    const handleEvent = jest.fn()
    render(
      <MenuContainer>
        <MenuItem
          icon={LucideRefreshCw}
          label="Sync Permissions"
          onClick={handleEvent}
        />
      </MenuContainer>,
    )

    fireEvent.click(screen.getByText('Sync Permissions'))
    expect(handleEvent).toHaveBeenCalledTimes(1)
  })

  it('renders Menu snapshots', () => {
    const { container } = render(
      <MenuContainer>
        <MenuItem href="." icon={LucideUserX} label="Revoke Admin" />
        <MenuItem
          icon={LucideRefreshCw}
          label="Sync Permissions"
          onClick={() => jest.fn()}
        />
        <MenuItem icon={LucideLogOut} label="Force Logout" />
        <MenuItem danger icon={LucideTrash2} label="Delete..." />
      </MenuContainer>,
    )

    expect(container).toMatchSnapshot()
  })
})
