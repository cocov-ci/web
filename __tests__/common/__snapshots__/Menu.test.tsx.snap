// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/Menu renders Menu snapshots 1`] = `
<div>
  <div
    class="base"
  >
    <div
      class="background"
    />
    <div
      class="content"
    >
      <a
        href="."
      >
        <div
          class="item"
        >
          <span
            class="label className"
          >
            Revoke Admin
          </span>
        </div>
      </a>
      <a
        href="javascript:void(0)"
      >
        <div
          class="item"
        >
          <span
            class="label className"
          >
            Sync Permissions
          </span>
        </div>
      </a>
      <div
        class="item"
      >
        <span
          class="label className"
        >
          Force Logout
        </span>
      </div>
      <div
        class="item danger"
      >
        <span
          class="label className"
        >
          Delete...
        </span>
      </div>
    </div>
  </div>
</div>
`;
